feat(devtools): work with stores added before app.use

marshallswain opened the pull request:
This removes the need to call `app.use(pinia)` BEFORE registering all services.  Services registered before calling the plugin are now held by a promise until the `app` object is set by the plugin.  The code used to set and get the `app` object for the devtools to work has been updated.  

As noted in the checkboxes, below, I haven't added any tests for this.  I did copy the built version into my Vite app to test. The app I'm testing has three stores that register before the app is created and five that register afterward.  Here's what I noticed from these changes:

- [x] Stores registered both before and after `app.use(pinia)` display properly in the devtools.
- [x] No extra logging occurs
- [x] No errors occur
- [x] Each store's data is browseable.

Summary of changes:
- convert getClientApp() to return a promise.  The promise resolves when the app is set (and a boolean ref gets set to true)
- Removes logging and warning state related to registering stores before calling app.use(pinia)

**What kind of change does this PR introduce?** (check at least one)

- [x] Feature
**Does this PR introduce a breaking change?** (check one)

- [x] No

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number) **not applicable**
- [x] All tests are passing

If adding a **new feature**, the PR's description includes:

- [x] A convincing reason for adding this feature: Nobody will ask you for this feature in the future.  ;)

marshallswain committed:
feat: Stores add to devtools before app.use(pinia)

Summary of changes:
- convert getClientApp() to return a promise.  The promise resolves when the app is set (and a boolean ref gets set to true)
- Removes logging and warning state related to registering stores before calling app.use(pinia)
- Updates the `addDevTools` function to return early if the store has already been registered (this prevents multiple “store installed” messages in the browser.

marshallswain commented:
@posva maybe I should have checked with you before making this PR.  Not sure if you've seen it, yet.  I know conference time is busier than normal.  Does this look like something worth merging?

posva commented on src/rootStore.ts:
```diff
@@ -53,8 +53,23 @@ export const storesMap = new WeakMap<
+  promise =
+    promise ||
+    new Promise((resolve) => {
+      watch(
```
rather than creating a watcher, I think we should keep the resolve function in a variable that is then called by `setClientApp()`

marshallswain replied:
Wow.  I've literally never used that pattern, but that's way easier.  Thanks!

posva commented on src/rootStore.ts:
```diff
@@ -53,8 +53,23 @@ export const storesMap = new WeakMap<
+}
+let promise: Promise<App> | undefined
+export const getClientApp = () => {
+  promise =
```
I think this should rather be

```js
return clientApp ? Promise.resolve(clientApp) : new Promise(...)
```

(also probably without the return)

posva reviewed (changes requested):
Thanks for this!

marshallswain committed:
fix(view): Remove watcher from getClientApp

As requested in feedback, this

- Removes the watcher from `getClientApp`
- Passes the promise’s resolve function to the outer scope.
- Calls that resolve function during `setClientApp`

marshallswain committed:
fix(view): must resolve the original promise

The `getClientApp` and `setClientApp` functions are now even cleaner, having removed the call to Promise.resolve.  This now assures that the original promise is resolved properly.

I was able to make it a bit shorter by taking advantage of the assignment operator returning the assigned value.

This also removes the export for clientApp, since the clientApp variable can be retrieved from the getClientApp promise.

marshallswain committed:
fix(view): update comment

marshallswain commented:
@posva With your suggestions I was able to make it even more succinct.  Note that I also removed the exported `clientApp` variable, since we can just pull it from the promise.  I noticed that the devtools still work without it, which I figured was the primary concern for removing it.

marshallswain commented:
I also took a look at the docs and it doesn't seem like anything needs to be updated.

marshallswain commented:
I also took a look at the docs and it doesn't seem like anything needs to be updated.

posva commented on src/rootStore.ts:
```diff
@@ -50,11 +50,14 @@ export const storesMap = new WeakMap<
-export let clientApp: App | undefined /*#__PURE__*/
-export const setClientApp = (app: App) => (clientApp = app)
-export const getClientApp = () => clientApp
+let promise: Promise<App> | undefined
```
```suggestion
let clientAppPromise: Promise<App> | undefined
```

posva commented on src/rootStore.ts:
```diff
@@ -50,11 +50,14 @@ export const storesMap = new WeakMap<
-export const setClientApp = (app: App) => (clientApp = app)
-export const getClientApp = () => clientApp
+let promise: Promise<App> | undefined
+let resolveApp: Function | undefined
```
```suggestion
let resolveApp: ((app: App) => void) | undefined
```

posva commented on src/rootStore.ts:
```diff
@@ -50,11 +50,14 @@ export const storesMap = new WeakMap<
+let promise: Promise<App> | undefined
+let resolveApp: Function | undefined
+export const setClientApp = (app: App) =>
+  resolveApp && (resolveApp as Function)(app)
```
with the change above the cast should be able to disappear
```suggestion
  resolveApp && resolveApp(app)
```

posva reviewed (changes requested):
Nice, looking good

marshallswain committed:
feat(view): clean up clientAppPromise in rootStore

Co-authored-by: Eduardo San Martin Morote <posva@users.noreply.github.com>

marshallswain commented:
@posva thanks for the feedback.  I've implemented the changes.

posva reviewed (approved):
Thanks!
